### PR TITLE
Enable bool tuple indexing

### PR DESCRIPTION
**Pull request by bradcray**
It's recently been pointed out that while Chapel supports indexing arrays by booleans (where the interpretation is to coerce to an integer and grab element 0 or 1), tuples traditionally have not.  I believe that this is a simple oversight rather than an intentional design choice, and this PR adds the ability to index tuples by boolean expressions.  A pair of tests demonstrate indexing tuples and arrays by booleans (both param and non-param values).

This feature is particularly useful now that we use 0-based indexing for tuples (and for arrays with inferred indexing), supporting the ability to write `("this", "that")(myBooleanExpr)` to choose between "this" and "that".  Note that unlike conditional expressions (`if myBooleanExpr then "that" else "this"`), the tuple/array forms evaluate both conditions rather than just the one corresponding to the bool.  This means that if you're selecting between side effecting functions `(foo(), bar())(myBooleanExpr)`, both functions will be called when using boolean indexing.

### CODE REVIEW

File: compiler/resolution/preFold.cpp [hunk 1]
```diff
@@ -2334,6 +2334,12 @@ static Expr* resolveTupleIndexing(CallExpr* call, Symbol* baseVar) {
       USR_FATAL_CONT(call, "tuple index %" PRIu64 " is out of bounds", uindex);
       error = true;
     }
+  } else if (get_bool(call->get(3), &uindex)) {
+    sprintf(field, "x%" PRIu64, uindex);
+    if (uindex >= (unsigned long)baseType->fields.length-1) {
+      USR_FATAL_CONT(call, "tuple index %" PRIu64 " is out of bounds", uindex);
+      error = true;
+    }
```
Comment (e-kayrakli) [inline compiler/resolution/preFold.cpp:2342]: Should this case be INT_FATAL? It looks like we can't have 0-sized tuples today, so having this fail can only mean we have messed up something in the compiler.

File: modules/internal/ChapelTuple.chpl [hunk 2]
```diff
@@ -216,6 +216,19 @@ module ChapelTuple {
...
 
+  pragma "no doc"
+  pragma "reference to const when const this"
+  pragma "star tuple accessor"
+  proc _tuple.this(i : bool) ref {
+    if !isHomogeneousTuple(this) then
+      compilerError("invalid access of non-homogeneous tuple by runtime value");
+    if boundsChecking then
+      if i < 0 || i > size-1 then
+        halt("tuple access out of bounds: ", i);
```
Comment (e-kayrakli) [inline modules/internal/ChapelTuple.chpl:227]: Similarly here. If we are sure that this is our error, maybe we should make this halt clearer

File: compiler/resolution/preFold.cpp (hunk 1 above)
Comment (bradcray) [reply compiler/resolution/preFold.cpp:2342]: We do have 1-tuples, though...  (and hope to have 0-tuples)

Comment (e-kayrakli) [reply compiler/resolution/preFold.cpp:2342]: Oops, right. I don't know what I was thinking.

On a similar vein, though, is it concerning to be able to bool-index a tuple that has more than 2 items?

Comment (bradcray) [reply compiler/resolution/preFold.cpp:2342]: I don't think so.  It seems as reasonable as it is to use an `int(8)` value to index into an array that has > 256 items.  It won't get you access to everything, but that doesn't make it unreasonable to do.

Comment (e-kayrakli) [reply compiler/resolution/preFold.cpp:2342]: That's fair.

**Review by e-kayrakli — APPROVED** [review]

**Pull request merged**